ipatests/test_integration/test_epn.py

```diff
@@ -674,3 +757,95 @@ def test_EPN_reinstall(self):
+            for i in range(users_per_day):
+                username = "user_with_pass_%d" % i
+                expiration = datetime_to_generalized_time(
+                    datetime.datetime.utcnow() + datetime.timedelta(
+                        days=max_days - day_i))
+                user = generate_user(self.master, username,
+                                     krbPasswordExpiration=expiration,
+                                     userPassword=user_password)
+                users.append(user)
+        # add the rest of users with password expiring in a week
+        for i in range(remainder):
+            username = "user_with_pass_%d" % i
+            expiration = datetime_to_generalized_time(
+                datetime.datetime.utcnow() + datetime.timedelta(days=7))
+            user = generate_user(self.master, username,
+                                 krbPasswordExpiration=expiration,
+                                 userPassword=user_password)
+            users.append(user)
+
+        # Users with wide unicode in usernames and with passwords to expire
+        expiration = datetime_to_generalized_time(
+            datetime.datetime.utcnow() + datetime.timedelta(days=7)
+        )
+        for i in range(users_in_group):
+            username = "%s_%d" % (''.join(
+                generate_random_unicode_string()), i)
+            user = generate_user(self.master, username,
+                                 krbPasswordExpiration=expiration,
+                                 userPassword=user_password)
+            users.append(user)
+
+        # Users with wide unicode in e-mail address
+        for i in range(users_in_group):
+            username = "user_with_unicode_name_%d" % i
+            email = "%s.%d@%s" % (''.join(
+                generate_random_unicode_string()), i, self.master.domain)
+            user = generate_user(self.master, username,
+                                 krbPasswordExpiration=expiration,
+                                 userPassword=user_password,
+                                 Mail2=email)
+            users.append(user)
+
+        # Users with multiple e-mail addresses
+        for i in range(users_in_group):
+            username = "user_multi_emails_%d" % i
+            email_1 = "%s_1@%s" % (username, self.master.domain)
+            email_2 = "%s_2@%s" % (username, self.master.domain)
+
+            user = generate_user(self.master, username,
+                                 krbPasswordExpiration=expiration,
+                                 userPassword=user_password,
+                                 Mail2=email_1,
+                                 Mail3=email_2)
+            users.append(user)
+
+        return users
+
+    def add_user_batch_to_ldap(self, users):
+        useradd_ldif = ""
+        for user in users:
+            # benchmarks show that + concat is the fastest for large strings
+            useradd_ldif += str(generate_user_ldif(user))
```

Comment:
Alternatively if you create all the users as a list and then loop over that list you can let LDIFWriter generate the huge string for you:

```
def generate_users_ldif(users):
    output = io.StringIO()
    writer = ldif.LDIFWriter(output, None, 76, '\n')
    for user_dict in users:
        user_dict_bytes = {
            key: val.encode('utf-8') if not isinstance(val, list)
            else [item.encode('utf-8') for item in val]
            for key, val in user_dict.items()
        }
        user_dict_bytes.pop('dn')
        writer.unparse(dn=user_dict['dn'], record=user_dict_bytes)
    output.seek(0)
    return output.read()
```